File: Algorithms/p94_Binary_Tree_Inorder_Traversal_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        vector<TreeNode*> stack;
        while(1){
            if(root){
                stack.push_back(root);
                root = root->left;
            }else{
                if(!stack.empty()){
                    res.push_back((stack.back())->val);
                    root = (stack.back())->right;
                    stack.pop_back();
                }else{
                    break;
                }
            }
        }
        return res;
    }
};

File: Algorithms/p94_Binary_Tree_Inorder_Traversal_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        while True:
            if root:
                stack.append(root)
                root = root.left
            else:
                if stack:
                    res.append(stack[-1].val)
                    root = stack[-1].right
                    stack.pop()
                else:
                    break
        return res
